### PR TITLE
Laboratory work #3, Nikita Gruzdev - 23FPL2

### PR DESCRIPTION
**Pull request by KolkiRik**

### CODE REVIEW

File: lab_3_ann_retriever/main.py [hunk 1]
```diff
@@ -130,6 +156,9 @@ def _remove_stop_words(self, tokens: list[str]) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not tokens or not isinstance(tokens, list):
+            return None
+        return list(word for word in tokens if not word in self._stop_words)
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:161]: can we create a list right away as list comprehension?

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -157,6 +190,22 @@ def build(self) -> bool:
...
         """
+        if not self._corpus or not isinstance(self._corpus,list):
+            return False
+        self._vocabulary = sorted(list({word for text in self._corpus for word in text}))
+        for word in self._vocabulary:
+            word_instances = 0
+            for text in self._corpus:
+                if word in text:
+                    word_instances += 1
+            self._idf_values[word] = math.log((len(self._corpus) - word_instances + 0.5) / (word_instances + 0.5))
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:201]: use `calculate_idf` from lab_2

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -183,6 +235,16 @@ def vector2tokens(self, vector: Vector) -> list[str] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not vector or len(self._token2ind) != len(vector):
+            return None
+        tokens = []
+        for pose, index in enumerate(vector):
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:241]: is it really need to have a loop by `vector`?

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -253,6 +326,15 @@ def index_documents(self, documents: list[str]) -> bool:
 
         In case of corrupt input arguments, False is returned.
         """
+        if not documents or not isinstance(documents, list) or not\
+                all(isinstance(word, str) for word in documents):
+            return False
+        self._documents = documents
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:332]: is it really need?

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great work! I'm waiting for the implementation on 8 score.

**Comment by Vasilisa-Blyudova** [comment]
No changes from the previous review

File: lab_3_ann_retriever/main.py (hunk 1 above)
Comment (KolkiRik) [reply lab_3_ann_retriever/main.py]: done!

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (KolkiRik) [reply lab_3_ann_retriever/main.py]: used

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (KolkiRik) [reply lab_3_ann_retriever/main.py]: I guess we can do it the opposite way using self vocab

File: lab_3_ann_retriever/main.py (hunk 4 above)
Comment (KolkiRik) [reply lab_3_ann_retriever/main.py:332]: I'm not sure what you mean

**Comment by KolkiRik** [comment]
oh no

**Comment by KolkiRik** [comment]
importing idf from 2nd lab broke it??

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (KolkiRik) [reply lab_3_ann_retriever/main.py]: importing idf from lab2 broke my start.py. Is it supposed to do that or did something go wrong on my part? removed one vectorizer.build() line from start and it seemed to fix it, but i'm not sure if that's what i'm supposed to do with that

**Comment by KolkiRik** [comment]
where did all these mypy problems come from?!

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -157,6 +193,21 @@ def build(self) -> bool:
         Returns:
             bool: True if built successfully, False in other case
         """
+        if not self._corpus or not isinstance(self._corpus,list):
+            return False
+        self._vocabulary = sorted(list({word for text in self._corpus for word in text}))
+        for word in self._vocabulary:
+            idf = calculate_idf(self._vocabulary, self._corpus)
+            if not idf:
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:201]: use `is None` instead of `not`

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -157,6 +193,21 @@ def build(self) -> bool:
...
+                return False
+            self._idf_values = idf
+            self._token2ind[word] = self._vocabulary.index(word)
+        if not self._vocabulary or not self._idf_values or not self._token2ind:
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:205]: .

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -269,6 +351,17 @@ def retrieve_relevant_documents(
 
         In case of corrupt input arguments, None is returned.
         """
+        if not query or not isinstance(query, str) or not isinstance(n_neighbours, int):
+            return None
+        queryVector = self._index_document(query)
+        if not queryVector:
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:357]: .

Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:357]: use snake case

File: lab_3_ann_retriever/main.py [hunk 8]
```diff
@@ -304,6 +397,15 @@ def retrieve_vectorized(self, query_vector: Vector) -> str | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not query_vector or not isinstance(query_vector,tuple):
+            return None
+        for vector in self._document_vectors:
+            if len(query_vector) != len(vector):
+                return None
+        doc_num = self._calculate_knn(query_vector, self._document_vectors, 1)
+        if not doc_num:
```
Comment (Vasilisa-Blyudova) [inline lab_3_ann_retriever/main.py:406]: .

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Very good work! Fix comments and tests and go to 10!

**Comment by KolkiRik** [comment]
Will start doing steps for 10 in a few hours (hopefully) and finish tomorrow (hopefully)

File: lab_3_ann_retriever/main.py [hunk 9]
```diff
@@ -157,6 +195,21 @@ def build(self) -> bool:
...
+        if not self._corpus or not isinstance(self._corpus,list):
+            return False
+        self._vocabulary = sorted(list({word for text in self._corpus for word in text}))
+        for word in self._vocabulary:
+            idf = calculate_idf(self._vocabulary, self._corpus)
+            if idf is None:
+                return False
+            self._idf_values = idf
+            self._token2ind[word] = self._vocabulary.index(word)
+        if self._vocabulary is None or self._idf_values is None or self._token2ind is None:
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/main.py:207]: `return None not in (self._vocabulary, self._idf_values, self._token2ind)`

File: lab_3_ann_retriever/start.py [hunk 10]
```diff
@@ -28,9 +30,47 @@ def main() -> None:
...
+    searchengine.index_documents(documents)
+    secret_tokens = text.split(', ')
+    secret_vector = tuple(float(token) for token in secret_tokens)
+    print(secret_vector)
+    secret = vectorizer.vector2tokens(secret_vector)
+    print(secret)
+    secret_revealed = searchengine.retrieve_vectorized(secret_vector)
+    print(secret_revealed)
+
+    searchengine_pro = SearchEngine(vectorizer, tokenizer)
```
Comment (artyomtugaryov) [inline lab_3_ann_retriever/start.py:71]: Try to find relevant document

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**